### DISCUSSION TITLE
How to use remotePatterns?

### DISCUSSION THREAD

**Discussion by devstefancho-inara** [vercel/next.js#61071] [Help] [answered]
### Summary

Hello Next.js Community,

I need to handle different subdomains for my domain (`my-domain.co.kr`).
I have three subdomains that I need to address

```
https://read.dev.my-domain.co.kr
https://read.beta.my-domain.co.kr
https://read.my-domain.co.kr
```

My question is about using wildcards in remotePatterns. Can I use a wildcard pattern like 

```javascript
// next.config.js

module.exports = {
// (...)
  images: {
    remotePartterns = [{
      protocol: 'https', 
      hostname: 'read.*.my-domain.co.kr'
    }]
  }
// (...)
```

Is this can match all three of these subdomains?
Any advice or examples of handling this situations would be greatly appreciated.
Thank you in advance for your help!

**Comment by icyJoseph** [comment, marked as answer]
Hi,

First of all, there's a typo on your configuration, `remotePartterns` should be `remotePatterns`.

Now, onto the good stuff. What does Next.js do with these? Or rather, how does it process them?

- https://github.com/vercel/next.js/blob/canary/packages/next/src/shared/lib/match-remote-pattern.ts#L34

We can see that it uses `picomatch`'s `makeRe`, within the `matchRemotePattern` function, which is, in turn, called by `hasMatch`.

I've rewritten these functions, without Type Annotations, and as close as possible to the current implementation.

```js
const { makeRe } = require("picomatch");

function matchRemotePattern(pattern, url) {
  if (pattern.protocol !== undefined) {
    const actualProto = url.protocol.slice(0, -1);
    if (pattern.protocol !== actualProto) {
      return false;
    }
  }
  if (pattern.port !== undefined) {
    if (pattern.port !== url.port) {
      return false;
    }
  }

  if (pattern.hostname === undefined) {
    throw new Error(
      `Pattern should define hostname but found\n${JSON.stringify(pattern)}`,
    );
  } else {
    if (!makeRe(pattern.hostname).test(url.hostname)) {
      return false;
    }
  }

  if (!makeRe(pattern.pathname ?? "**").test(url.pathname)) {
    return false;
  }

  return true;
}

function hasMatch(domains, remotePatterns, url) {
  return (
    domains.some((domain) => url.hostname === domain) ||
    remotePatterns.some((p) => matchRemotePattern(p, url))
  );
}

const remotePatterns = [
  {
    protocol: "https",
    hostname: "read(.*)?.my-domain.co.kr",
  },
];

const urls = [
  "https://read.dev.my-domain.co.kr",
  "https://read.beta.my-domain.co.kr",
  "https://read.my-domain.co.kr",
];

urls.forEach((path) => {
  console.log(hasMatch([], remotePatterns, new URL(path)));
});

```

I named this `index.js`. Then I did `yarn init -y`,  installed with `yarn add picomatch`, and then `node index.js`

I got

```console
true
true
true
```

However, using yours: `read.*.my-domain.co.kr`, I get `true, true, false`.

Alright, so the pattern you kind of wanted seems to be `"read(.*)?.my-domain.co.kr"`.

However, I'd argue against this. I'd say it is better to be explicit, rather than buying into a regex based solution. Also, domains such as `"read.evil.my-domain.co.kr"`, would work, which is not what you wanted.

Sure you can limit the group capture to `(dev | beta)`, but that's just adding complexity.

Why not just do:

```js
const hostnames = [
  'read.dev.my-domain.co.kr',
  'read.beta.my-domain.co.kr',
  'read.my-domain.co.kr',
];

const remotePatterns = hostnames.map((hostname) => ({
  protocol: 'https',
  hostname,
}));
```

That makes it explicit, and easy to reason about:

```js

const hostnames = [
  'read.dev.my-domain.co.kr',
  'read.beta.my-domain.co.kr',
  'read.my-domain.co.kr',
];

module.exports = {
  images: {
    remotePatterns: hostnames.map((hostname) => ({
      protocol: 'https',
      hostname,
    })),
  },
};
```

**Reply by devstefancho-inara** [reply]
@icyJoseph Thank you for providing actual code example, It's really clarified things for me

**Comment by devstefancho-inara** [comment]
@icyJoseph The answer is marked but I have one more curious thing.

In nextjs documentation `**` wildcard may not work in the middle of the pattern

![image](https://github.com/vercel/next.js/assets/142373285/e676552d-5f05-4eaf-b8a4-b95894b85b25)


but when I try this pattern(`"read.**.my-domain.co.kr"`)
The result is like this
```
const urls = [
  "https://read.dev.my-domain.co.kr",            // true
  "https://read.beta.my-domain.co.kr",          // true
  "https://read.my-domain.co.kr",                  // false
  "https://read.more-sub.dev.my-domain.co.kr", // true
];
```

I think, last one should be false. but it isn't
Could you explain this too?
Thank you

**Reply by icyJoseph** [reply]
Mmm I am not sure about that. This is where picomatch's glob rules start to apply. 

- https://github.com/micromatch/picomatch?tab=readme-ov-file#basic-globbing
